src/main/java/de/tum/in/www1/artemis/service/ProgrammingSubmissionService.java

```diff
@@ -126,4 +133,29 @@ public ProgrammingSubmission getLatestPendingSubmission(Long participationId) th
         }
         return submissionOpt.get();
     }
+
+    /**
+     * Create a submission with type INSTRUCTOR for the last commit hash of the given participation.
+     *
+     * @param participation to create submission for.
+     * @param submissionType of the submission to create.
+     * @return created submission.
+     * @throws IllegalStateException if the last commit hash can't be retrieved.
+     */
+    @Transactional
+    public ProgrammingSubmission createSubmissionForParticipation(ProgrammingExerciseParticipation participation, SubmissionType submissionType) throws IllegalStateException {
+        URL repoUrl = participation.getRepositoryUrlAsUrl();
+        ObjectId lastCommitHash;
+        try {
+            lastCommitHash = gitService.getLastCommitHash(repoUrl);
```

Comment:
This basically means that we would have 2 programming submissions with the same commit hash, right?
When the result from Bamboo is retrieved, we search for the programming submission with a specific commit hash. What happens then in case we find two submission with the same commit hash?

Are you sure that there can be no problems in the ResultService / BambooService in such a case?

Reply:
I think there is no other way to handle it, if we don't want to make an empty commit.

The Bamboo.onBuildCompletedNew fetches the pending submissions ordered by submission date desc and then tries to find the first one with a matching commit hash:

```
            List<ProgrammingSubmission> submissions = programmingSubmissionRepository.findByParticipationIdAndResultIsNullOrderBySubmissionDateDesc(participation.getId());
            Optional<ProgrammingSubmission> latestMatchingPendingSubmission = submissions.stream().filter(submission -> {
                String matchingCommitHashInBuildMap = getCommitHash(buildMap, submission.getType());
                return matchingCommitHashInBuildMap.equals(submission.getCommitHash());
            }).findFirst();
```

So the "worst case scenario" would be:
- Student executes a submission, the build is created on Bamboo
- The build takes longer than 2 minutes, this enables the student to trigger the submission again
- A new submission with the same submission is created on the server, there are now 2 submissions for the same commitHash and 2 running builds
- The first build returns a result to Artemis, this result is now attached to the second submission (that was just created)
- The second build finishes and returns a result to Artemis, this result is attached to the first submission

Reply:
(I also renamed the method to `createSubmissionWithLastCommitHashForParticipation` to make the intention more clear.

Reply:
I think this is fine, as long as we do not get severe problems by this approach. The worst case scenario sounds ok to me, but I would suggest that you add this as comment into the source code. I would write it as detailed as possible so that future developers are aware of it

Reply:
You are right, I changed to JavaDoc:

```
    /**
     * Create a submission with given submission type for the last commit hash of the given participation.
     * WARNING: The commitHash is used to map incoming results to submissions. Using this method could cause the result to have multiple fitting submissions.
     *
     * @param participation to create submission for.
     * @param submissionType of the submission to create.
     * @return created submission.
     * @throws IllegalStateException if the last commit hash can't be retrieved.
     */
```

Reply:
Can you put the whole comment of this thread into the JavaDoc as discussion? even the worst case scenario?

Reply:
I added a link to this conversion and the worst case scenario to the docs.